Babelstream v5.0 Spack Package Updates

kaanolgu opened the pull request:
This update makes the Spack package compatible with the latest version 5.0 of Babelstream alongside with added support for Fortran packages!

kaanolgu commented:
@spackbot fix style

tomdeakin commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -15,27 +19,34 @@ def find_model_flag(str):
+    version("3.0", sha256="776219c72e0fdc36f134e6975b68c7ab25f38206f8f8af84a6f9630648c24800")
+    version("1.0", sha256="3cfb9e45601f1f249878355c72baa6e6a61f6c811f8716d60b83c7fb544e1d5c")
+    # version("main", branch="main")
     maintainers("tomdeakin", "kaanolgu", "tom91136", "robj0nes")
```
```suggestion
    maintainers("tomdeakin", "kaanolgu", "tom91136")
    # Previous maintainers: "robj0nes"
```

tomdeakin reviewed (approved):
Just one minor change to update the maintainers.

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -15,27 +19,35 @@ def find_model_flag(str):
     # Also supported variants are cuda and rocm (for HIP)
     variant("sycl", default=False, description="Enable SYCL support")
-    variant("sycl2020", default=False, description="Enable SYCL support")
+    variant("sycl2020usm", default=False, description="Enable SYCL support")
+    variant("sycl2020acc", default=False, description="Enable SYCL support")
```
These variants could use somewhat more description to say how they differ.

Also, would it be better to have a single `sycl` variant and another variant that is only relevant when `+sycl`?

kaanolgu replied:
According to your comments down below and here, it seems like it would be better to make a single `sycl` variant with a new variant that might be an `option` to choose from ? and actually, `option` could replace the `implementation` variant too since you mentioned in the later comments that it is not ideal.

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -70,7 +82,10 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
+        "offload",
+        values=str,
+        default="none",
+        description="Enable RAJA Target [CPU or NVIDIA] / Offload with custom settings for OpenMP",
```
Is the offload still NVIDIA specific? Is `[CPU or NVIDIA]` specifying the valid values?

kaanolgu replied:
Yes, from the CMake file it was : 

https://github.com/UoB-HPC/BabelStream/blob/2f00dfb7f8b7cfe8c53d20d5c770bccbf8673440/src/raja/model.cmake#L19

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -105,7 +122,8 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
     )
-
+    # OpenMP Intel Offload
+    variant("intel_target", values=str, default="none", description="Intel Offload target")
```
Should this logic be unified with the `offload` variant above?

kaanolgu replied:
Thank you for spotting this, will merge them. One quick question would be whether using fewer `variant` names and using them in multiple models is favoured. Since we have 11+ models, it is hard to reduce the number of `variant`s ?

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -140,38 +158,67 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
-    )
+    with when("build_system=cmake"):
+        variant(
+            "implementation",
```
This name seems less specific than ideal

kaanolgu replied:
Since some descriptions were referred to as `implementation` in the original CMake package, having a similar naming trend could be easier to follow. However, brainstorming a new variant name is an option.

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -140,38 +158,67 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
+            "Sequential",
+        ]
+        variant(
+            "impl", values=implementation_vals, default="Sequential", description="Implementation"
```
Is this fundamentally the same `implementation` variant as above, or something else? It's not clear what this is doing from the code

kaanolgu replied:
The `implementation` is used for the `CMake` related models, and `impl` is used for the `makefile` related (Fortran) models. Could introducing a distinction between these be beneficial for the users or not?

becker33 reviewed (changes requested):
Not a full review, I didn't check all of the code for implementing the variants/options. Just looking at the structure of the variants so far.

kaanolgu committed:
Addressing comments made by becker33:

- Merging sycl2020usm and sycl2020acc into sycl2020 and the submodel=acc/usm variant is introduced
- implementation is renamed to option
- impl ( fortran implementation options) renamed to foption
- intel_target variant is renamed as cpu_target and used in other models that require cpu_target to be specified

douglasjacobsen commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -84,14 +121,17 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("thrust", when="+thrust")
+    depends_on("cuda", when="+thrust")
+    depends_on("hip", when="+hip")
     depends_on("rocthrust", when="+thrust implementation=rocm")
```
I think this dependency needs to be updated now. It looks like the `implementation` variant was removed, so now this dependency is broken.

kaanolgu replied:
Thank you for spotting this Doug, will push a commit to fix it asap!

kaanolgu committed:
Fix for rocthrust dependency condition naming from implementation to option

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -45,7 +56,30 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
     # this variant enables a path to be provided.
     variant("dir", values=str, default="none", description="Enable Directory support")
-
+    variant("submodel", values=str, default="none", description="SYCL2020 -> choose between usm and acc methods")
```
```suggestion
    variant("submodel", values=("usm", "acc"), when="+sycl2020", default="usm", description="SYCL2020 -> choose between usm and acc methods")
```

becker33 replied:
Then we can delete the conflicts statement below that implements this when clause.

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -70,7 +104,10 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
     variant(
-        "offload", values=str, default="none", description="Enable RAJA Target [CPU or NVIDIA]"
+        "offload",
+        values=str,
```
```suggestion
        values=("cpu", "nvidia"),
```

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -45,7 +56,30 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
     variant("dir", values=str, default="none", description="Enable Directory support")
-
+    variant("submodel", values=str, default="none", description="SYCL2020 -> choose between usm and acc methods")
+    variant("option", values=str, default="none", description="For SYCL/SYCL2020 choose compiler implementation option, for THRUST which THRUST implementation to use [CUDA/ROCM]")
```
This should be implemented as two separate conditional variants, something like

```
    variant("thrust_backend", values=("cuda", "rocm"), default="cuda", when="+thrust", description="thrust backend implementation to use")

    sycl_compiler_implementations = []  # list them here, with the default first
    variant("sycl_compiler_implementation", values=sycl_compiler_implementations, default=sycl_compiler_implementations[0], when="+sycl", description="...")
    variant("sycl_compiler_implementation", values=sycl_compiler_implementations, default=sycl_compiler_implementations[0], when="+sycl2020", description="...")

becker33 reviewed (changes requested):
Please try to make the variant setup less convoluted. We should not have a single variant spanning two very different and unrelated semantics based on the values of the other variants.

kaanolgu committed:
Apply suggestions from code review of becker33

Co-authored-by: Greg Becker <becker33@llnl.gov>

kaanolgu committed:
Splitting option variant into sycl_compiler_implementation and thrust_backend

kaanolgu commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -102,17 +96,13 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage, MakefilePackage):
+    # Raja offload
     variant(
         "offload",
         values=("cpu", "nvidia"),
```
Offload is shared between others too like `std-data` etc. I will push a new commit to address those too

kaanolgu committed:
reducing stddata,stdindices,stdranges to a single std with std_submodel introduction

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -68,14 +92,28 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
+    variant(
+        "std_use_tbb",
+        values=("on","off"),
+        default="off",
```
Why not use a boolean variant `(True, False)` here?

kaanolgu replied:
The issue was we were checking the `models` with boolean ( cuda, acc, omp etc. ) but since you mentioned this I thought it would be better to create a list of models and neglect anything except those models. So in the new commit it is now boolean and the `-DMODEL` filtering is more precise and less crowded to handle different scenarios

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -182,40 +234,108 @@ def cmake_args(self):
+            and ("~acc" in self.spec)
+            and ("~omp" in self.spec)
+            and ("~thrust" in self.spec)
+        ):
```
```suggestion
        if self.spec.satisfies("+cuda~kokkos~acc~omp~thrust"):
```

kaanolgu replied:
Thanks, I have updated all the `self.spec` checks should be less crowded now

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -234,37 +356,61 @@ def cmake_args(self):
-                cuda_arch = "sm_" + cuda_arch_list[0]
-                args.append("-DOFFLOAD= " + "NVIDIA:" + cuda_arch)
-            elif "amdgpu_target" in self.spec.variants:
+                cuda_arch = "cc" + cuda_arch_list[0]
```
If the change from `sm_` to `cc` is intentional, please update the comment above.

kaanolgu replied:
I have updated the comments, the logic is : 
```
any model with  : -DCUDA_ARCH = sm_XY
any model with  : -arch = cc_XY 
(special ) since thrust has CMAKE_CUDA_ARCHITECTURES, it requires just value
```

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -234,37 +356,61 @@ def cmake_args(self):
+                target_device = "gpu" + ";"
+                cuda_arch_list = self.spec.variants["cuda_arch"].value
+                # the architecture value is only number so append sm_ to the name
+                cuda_arch = "cc" + cuda_arch_list[0]
```
These two lines appear duplicated

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -319,20 +469,22 @@ def cmake_args(self):
         # ===================================
         if "+raja" in self.spec:
-            args.append("-DCMAKE_CXX_COMPILER=" + self.compiler.cxx)
+            args.append("-DCMAKE_CXX_COMPILER=" + spack_cxx)
```
This seems to be set for every combination -- could we separate it out and actually set it unconditionally or am I missing something?

kaanolgu replied:
For the `+hip` it was : 
```
        if "+hip" in self.spec:
            hip_comp = self.spec["hip"].prefix + "/bin/hipcc"
            args.append("-DCMAKE_CXX_COMPILER=" + hip_comp)
```
It would be better to set this to `spack_cxx` and let the user install `hip` and use it with spack install babelstream%hip ? 
For a bit of brainstorming it would be better actually since the user could decide to test different versions of compiler etc. 
Also the `+thrust` and `+tbb` does not need `CMAKE_CXX_COMPILER` to be set so created a special if condition for that.
Thank you for pointing out this

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -414,3 +572,232 @@ def cmake_args(self):
+        if spec.compiler.name == "arm":
+            flags = "-std=f2018 " + pkg.compiler.opt_flags[4] + " -Wall -Wno-unused-variable"
+
+            if platform.machine() == "aarch64":
+                # TODO: Add MCPU option here as in -mcpu=$(MCPU)
+                flags += "-mcpu=native"
+            else:
+                # TODO: Add MARCH option here as in -march=$(MARCH)
+                flags += "-march=native"
```
These flags are going to conflict with the flags from `archspec` that Spack will be emitting.

Is there a reason these need to be set to `native` rather than the values associated with the specific architecture targetted?

kaanolgu replied:
In the new commit I tried to address this and simplify it but I am getting `error`: 

```
$ spack install babelstream%gcc@13.1.0 build_system=makefile foption=DoConcurrent target=znver3
==> Error: concretization failed for the following reasons:

   1. 'babelstream target=broadwell' cannot satisfy constraint 'target=znver3'
```

becker33 replied:
Can you run `spack config blame packages | grep "broadwell"`? You shouldn't be stuck building for broadwell, and we definitely shouldn't say it's a broadwell and then throw non-broadwell flags.

kaanolgu replied:
I think it was the issue with MACS system of Isambard, sincet then I moved to check with Phase3 system (newer)  and didn't face this issue after that

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -414,3 +572,232 @@ def cmake_args(self):
+                "OpenACC",
+                "OpenACCArray",
+            ]:
+                sys.exit(self.spec.variants["foption"].value + "is not supported by this compiler")
```
Probably better to raise an error than to call `sys.exit` directly. Looks like most packages are using `InstallError` for similar issues.

kaanolgu replied:
Thank you! I tried to also simplify checking process which reduced number of lines too

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -414,3 +572,232 @@ def cmake_args(self):
+
+            # config['MARCH'] = "neoverse-v1,neoverse-n1,icelake-server,znver3,cortex-a78ae"
+
+            if platform.machine() == "aarch64":
+                # TODO: Add MCPU option here as in -mcpu=$(MCPU)
+                flags += "-mcpu=native"
+            else:
+                # TODO: Add MARCH option here as in -march=$(MARCH)
+                flags += "-march=native"
```
Same as before

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -414,3 +572,232 @@ def cmake_args(self):
+            # march=neoverse-v1,neoverse-n1,zen3
+            march = "none"
+            if march != "none":
+                if platform.machine() == "aarch64":
```
We should be keying off `spec.target` here, not `platform.machine()`.

`spec.target.family` gives the information at the granularity of `aarch64/x86_64/ppc64le`.

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -414,3 +572,232 @@ def cmake_args(self):
+                "OpenACC",
+                "OpenACCArray",
+            ]:
+                sys.exit(self.spec.variants["foption"].value + "is not supported by this compiler")
```
raise InstallError

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -414,3 +572,232 @@ def cmake_args(self):
+
+            # Error Messages
+            if self.spec.variants["foption"].value in ["CUDA", "CUDAKernels", "OpenACC"]:
+                sys.exit(self.spec.variants["foption"].value + "is not supported by this compiler")
```
raise InstallError

kaanolgu committed:
Modified std_use_tbb to be boolean; also changed model filtering algorithm to make sure that it only picks model names
Modified checking self.spec condition for each model to simplify with self.spec.satisfies
Modified comments to clear confusion with cuda_arch cc_ and sm_ prefix appends
Deleted duplicate of cuda_arch definition from +omp
CMAKE_CXX_COMPILER moved to be shared arg between all models except tbb and thrust
Replaced sys.exit with InstallError and created a dictionary to simplify things and eliminate excess code lines doing same checks
Replaced the -mcpu flags to -march since it is deprecated now
Replaced platform.machine with spec.target

kaanolgu committed:
Bug fixes

kaanolgu committed:
Further bug fixes and cleaning in code for all models

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -53,50 +80,103 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
         msg="RAJA implementation requires architecture to be specfied by dir=",
     )
-
+    # depends_on("cuda@12.3.0:", when="+thrust")
```
This looks like it's leftover from testing.

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -53,50 +80,103 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
-    depends_on("rocthrust", when="+thrust implementation=rocm")
-
+    depends_on("cuda", when="+thrust")
+    depends_on("cuda", when="+raja")
```
```suggestion
    depends_on("cuda", when="+raja raja_offload=NVIDIA")
```
Is this a direct dependency, or is it handled entirely through raja? My first thought is that it should be `depends_on("raja+cuda", ...)` and it's on raja to handle the cuda dep.

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -53,50 +80,103 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("thrust", when="+thrust")
-    depends_on("rocthrust", when="+thrust implementation=rocm")
-
+    depends_on("cuda", when="+thrust")
```
```suggestion
    depends_on("cuda", when="+thrust thrust_submodel=cuda"")
```
I'm not as familiar with thrust, but I'm assuming here that it only requires a dependency on cuda when targeting cuda? Also, the `+thrust` in the suggestion is not strictly necessary, as it's implied by `thrust_submodel=cuda`.

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -123,108 +203,225 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
+        msg="OpenCL implementation requires backend to be specfied by ocl_backend=",
     )
+    variant(
+        "raja_backend",
```
Why are there separate variants for `raja_offload` and `raja_backend`?

kaanolgu replied:
Thank you spotting this, I will merge them to be `raja_offload`

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -273,123 +481,125 @@ def cmake_args(self):
         if "+ocl" in self.spec:
-            if "backend" in self.spec.variants:
-                if "cuda" in self.spec.variants["backend"].value:
+            if "ocl_backend" in self.spec.variants:
```
Doesn't the structure of the variants guarantee this is True if the line above is?

kaanolgu replied:
Yes, thank you for spotting, I found a few more cases like this I got rid of them too

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -273,123 +481,125 @@ def cmake_args(self):
                 # if("cpu" in self.spec.variants['offload'].value):
-
-            if "omp" in self.spec.variants["backend"].value:
+            if "omp" in self.spec.variants["raja_backend"].value:
```
Looks like we're conflating `raja_backend` and `raja_offload`?

kaanolgu committed:
Removing raja_backend, introducing openmp_flag,removing -march flags,clearing debugging print(), removing excess if ___ in self.spec.variants

kaanolgu committed:
Merge branch 'develop' into develop

kaanolgu committed:
Merge branch 'develop' into develop

kaanolgu committed:
Merge pull request #2 from spack/develop

Update the Fork

kaanolgu committed:
Merge pull request #4 from spack/develop

dakota: make python dependency optional, add v6.19 (#42914)

kaanolgu committed:
Merge branch 'develop' into develop

kaanolgu committed:
Fixing style and audit error with raja+cuda

kaanolgu committed:
Fixing further style issues

kaanolgu commented:
@spackbot fix style

kaanolgu committed:
Reverting back some style changes which on my system spack style complained but on CI it's OK

kaanolgu committed:
deleting a single space

kaanolgu committed:
[FIX] Issue where Thrust couldn't find correct compiler (it requires nvcc)

kaanolgu committed:
[FIX] Fortran unsupported check to match the full string

kaanolgu committed:
[FIX] RAJA cuda_arch to be with sm_ not cc_

kaanolgu committed:
Merge branch 'develop' into develop

Commit:
Kokkos FIX
- dir= option is no longer needed for kokkos

Commit:
RAJA fix
- dir is no longer needed

tgamblin committed:
fix style

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -15,21 +17,26 @@ def find_model_flag(str):
+    version("3.1", sha256="be69e6085e8966e12aa2df897eea6254b172e5adfa03de0adbb89bc3065f4fbe")
+    version("3.0", sha256="776219c72e0fdc36f134e6975b68c7ab25f38206f8f8af84a6f9630648c24800")
+    version("1.0", sha256="3cfb9e45601f1f249878355c72baa6e6a61f6c811f8716d60b83c7fb544e1d5c")
+    # version("main", branch="main")
```
Do we need this?  why comment it out?

tomdeakin replied:
```suggestion
    version("main", branch="main")
    version("develop", branch="develop")
```

tgamblin commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -15,21 +17,26 @@ def find_model_flag(str):
+    version("1.0", sha256="3cfb9e45601f1f249878355c72baa6e6a61f6c811f8716d60b83c7fb544e1d5c")
+    # version("main", branch="main")
+    maintainers("tomdeakin", "kaanolgu", "tom91136")
+    # Previous maintainers: "robj0nes"
```
I don't think we need to list this.

tgamblin commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -38,67 +45,184 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
+        default="data",
+        description="STD -> choose between data, indices and ranges models",
+    )
+    variant(
+        "sycl_opencl_path",
+        values=str,
+        when="+sycl",
+        default="none",
+        description="[ComputeCpp only] Path to OpenCL library, usually called libOpenCL.so",
+    )
```
This is essentially an external but it's modeled as a variant -- it should be a dependency on the `opencl` virtual so that the user can choose different opencl implementations.  Look at some of the OpenCL providers:

```console
> spack providers opencl
opencl:
cuda@:6  ocl-icd@2.2.3:2.2.7   ocl-icd@2.2.12   opencl-icd-loader  pocl@3:
cuda@7:  ocl-icd@2.2.8:2.2.11  ocl-icd@2.2.13:  pocl
```
and dependents:
```console
> spack dependents opencl
arrayfire  clinfo                      gdal        kassiopeia  opencv
bohrium    cmdstan                     gearshifft  lammps      py-pyopencl
clblast    dbcsr                       grass       libbeagle   sgpp
clfft      double-batched-fft-library  hpx5        onednn      specfem3d-globe
```

tomdeakin replied:
Can probably just delete as ComputeCpp is gone

tgamblin commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -38,67 +45,184 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
-    # Kokkos conflict and variant
-    conflicts(
-        "dir=none", when="+kokkos", msg="KOKKKOS requires architecture to be specfied by dir="
+    sycl_compiler_implementations = [
+        "oneapi-icpx",
+        "oneapi-clang",
+        "dpcpp",
+        "hipsycl",
+        "computecpp",
+        "none",
+    ]  # list them here, with the default first
```
This is similarly something that should really be modeled as a virtual dependency... I am curious how @becker33 would recommend modeling this one as we transition to compiler deps -- there is already a `sycl` virtual provided by `intel-oneapi-runtime` and `hipsycl`, so that would be my preferred way.

For some of the others I think you should add packages to Spack that provide `sycl`, e.g. `computecpp`. `dpcpp` isn't supported anymore (Intel folks deprecated it) so you might consider leaving it out.

tomdeakin replied:
ComputeCpp is deprecated/gone so no need for a Spack package.

tgamblin commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -38,67 +45,184 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
+            - CUDA (via https://github.com/NVIDIA/thrust)\
+            - ROCM (via https://github.com/ROCmSoftwarePlatform/rocThrust)",
     )
+    variant(
+        "thrust_backend",
+        values=("cuda", "omp", "tbb"),
+        default="cuda",
+        when="+thrust",
+        description="Which THRUST implementation to use, supported options include option",
+    )
```
I *think* this should be a virtual as well, e.g. `thrust-api` so the user can pick their thrust implementation. This and other suggestions above would simplify all the conditional logic in your dependencies above.

kaanolgu replied:
@tgamblin  it just sets `args.append("-DBACKEND=" + self.spec.variants["thrust_backend"].value.upper())` so I think it is simple that doesn't need any extra effort on spack package ( taken care by the CMake)

Commit:
[omp] Adding clang support for nvidia offload

kaanolgu commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -475,33 +493,29 @@ def cmake_args(self):
+                        offload_args=" -mp=gpu;" + "-gpu=" + cuda_arch + " "
+                    if self.spec.satisfies("%clang"):
+                        cuda_arch = "sm_" + self.spec.variants["cuda_arch"].value[0]
+                        offload_args="-march=znver3;-fopenmp;--offload-arch="+cuda_arch
```
From CMake this was supplied `-march=native` but seems like the package didn't like this, how could we force `march` to be the CPU arch we have in the system ?  @tgamblin @becker33

tomdeakin commented:
@tgamblin @becker33  what's left to do to get this merged?

Commit:
SYCL2020 offload to nvidia GPU

Commit:
Fix for the compiler names

Commit:
removing deprecated compilers from sycl and adding offload to sycl2020

kaanolgu commented:
@tgamblin @becker33 @tomdeakin  addressed all your comments and addressed @rfbgo 's  feedback about various scenarios where it breaks or not functioning properly, I believe it is close to merge ( Thank you everyone! )

only one left is this : https://github.com/spack/spack/pull/41019#discussion_r1698833315 is it OK to modify this to provide arch from the package ?

kaanolgu committed:
Merge branch 'develop' into develop

kaanolgu commented:
@spackbot fix style

kaanolgu committed:
[@spackbot] updating style on behalf of kaanolgu

kaanolgu committed:
Merge branch 'develop' into develop

kaanolgu commented:
@spackbot fix style

kaanolgu commented:
Hi @tgamblin  and @becker33 , 
Are there any additional improvements you think should be made to this PR? If not, I will address the remaining style issues and then kindly ask for your approval, as this is necessary for the latest version of BabelStream due to the many new features and updates included. We can always make smaller PRs later to address any necessary fixes :) 

Thank you all once again for your help and support!

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -15,21 +17,26 @@ def find_model_flag(str):
-
-    maintainers("tomdeakin", "kaanolgu", "tom91136", "robj0nes")
-
+    build_system("cmake", "makefile", default="cmake")
```
Do all versions support both build systems?

kaanolgu replied:
No, the Fortran ones support Makefile build system and C++ ones support CMake build system

becker33 replied:
What does the sequencing look like for a full build of babelstream with each build system?

kaanolgu replied:
Each build system is separate and each model builds separately so we don't have any sequence for building

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -38,67 +45,170 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
+        description="No-op if ONE_TBB_DIR is set. Link against an in-tree oneTBB\
+                    via FetchContent_Declare, see top level CMakeLists.txt for details",
+    )
+    variant(
+        "std_use_onedpl",
+        values=(True, False),
+        default=False,
+        when="+std",
+        description="Link oneDPL which implements C++17 executor policies\
+                    (via execution_policy_tag) for different backends",
+    )
```
Couldn't this just be keyed off `std_use_onedpl=none -> False` and any other values -> `True`?

kaanolgu replied:
The aim here was if the user provides `+std_use_onedpl` it will be enabled if not it will be disabled. Also, limits user from inputting `std_use_onedpl=[2024,OK] ` etc. and enables us not to use conflicts which was suggested in the earlier set of reviews actually.  The none method would be similar to this ?  (replacing false with none ?) 

https://github.com/spack/spack/blob/d70e9e131d12d9c1ee9f155ecb44684999d1f593/var/spack/repos/builtin/packages/itk/package.py#L42-L45

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -108,125 +218,244 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
+    depends_on("intel-oneapi-compilers", when="+ocl ocl_backend=intel")
+    depends_on("pocl@1.5", when="+ocl ocl_backend=pocl")
+
+    variant("flags", values=str, default="none", description="Additional CXX flags to be provided")
```
Why not just use the `cxxflags` option on the spec for this? It seems redundant.

kaanolgu replied:
Actually, great attention to detail, it should be named as `extra_cuda_flags` since we are using it for providing extra cuda flags if needed adding it to next commit

EDIT: and have `fortran_flags`

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -108,125 +218,244 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
         spec_string = str(self.spec)
 
         # take only the first portion of the spec until space
         spec_string_truncate = spec_string.split(" ", 1)[0]
```
This is not a stable way to partition spec information -- what information are you trying to get here?

kaanolgu replied:
So I am trying to get `+cuda+thrust` etc. so the model names that set in a list, is there a way to do this ? 

Because for example user sets +thrust but has cuda backend and +cuda is also set automatically but we need to track this. but obviously not unrelated to our model choice so an example :

For `spack install babelstream@5.0%gcc +thrust cuda_arch=80 thrust_backend=cuda`
```
original: babelstream@=5.0%gcc@=8.5.0~acc+cuda~hip~ipo~kokkos~ocl~omp~raja~rocm~std~sycl~sycl2020~tbb+thrust build_system=cmake build_type=Release cpu_arch=none cuda_arch=80 cuda_memory_mode=default dir=none flags=none generator=make thrust_backend=cuda thrust_submodel=cuda arch=linux-rhel8-zen ^[deptypes=build,link] berkeley-db@=18.1.40%gcc@=8.5.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-rhel8-zen ^[deptypes=build,link] bzip2@=1.0.8%gcc@=8.5.0~debug~pic+shared build_system=generic arch=linux-rhel8-zen ^[deptypes=build] ca-certificates-mozilla@=2023-05-30%gcc@=8.5.0 build_system=generic arch=linux-rhel8-zen ^[deptypes=build] cmake@=3.30.2%gcc@=8.5.0~doc+ncurses+ownlibs build_system=generic build_type=Release patches=dbc3892 arch=linux-rhel8-zen ^[deptypes=build,link] cuda@=12.6.1%gcc@=8.5.0~allow-unsupported-compilers~dev build_system=generic arch=linux-rhel8-zen ^[deptypes=build,link] curl@=8.8.0%gcc@=8.5.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-rhel8-zen ^[deptypes=build] diffutils@=3.10%gcc@=8.5.0 build_system=autotools arch=linux-rhel8-zen ^[deptypes=link] gcc-runtime@=8.5.0%gcc@=8.5.0 build_system=generic arch=linux-rhel8-zen ^[deptypes=build,link] gdbm@=1.23%gcc@=8.5.0 build_system=autotools arch=linux-rhel8-zen ^[deptypes=build,link virtuals=iconv,libc] glibc@=2.28%gcc@=8.5.0 build_system=autotools arch=linux-rhel8-zen ^[deptypes=build,link] gmake@=4.4.1%gcc@=8.5.0~guile build_system=generic arch=linux-rhel8-zen ^[deptypes=build,link] libxml2@=2.10.3%gcc@=8.5.0+pic~python+shared build_system=autotools arch=linux-rhel8-zen ^[deptypes=build,link] ncurses@=6.5%gcc@=8.5.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-rhel8-zen ^[deptypes=build,link] nghttp2@=1.63.0%gcc@=8.5.0 build_system=autotools arch=linux-rhel8-zen ^[deptypes=build,link] openssl@=3.3.1%gcc@=8.5.0~docs+shared build_system=generic certs=mozilla arch=linux-rhel8-zen ^[deptypes=build] perl@=5.40.0%gcc@=8.5.0+cpanm+opcode+open+shared+threads build_system=generic arch=linux-rhel8-zen ^[deptypes=build virtuals=pkgconfig] pkgconf@=2.2.0%gcc@=8.5.0 build_system=autotools arch=linux-rhel8-zen ^[deptypes=build,link] readline@=8.2%gcc@=8.5.0 build_system=autotools patches=bbf97f1 arch=linux-rhel8-zen ^[deptypes=build,link] thrust@=1.17.2%gcc@=8.5.0~ipo build_system=cmake build_type=Release generator=make arch=linux-rhel8-zen ^[deptypes=build,link] xz@=5.4.6%gcc@=8.5.0~pic build_system=autotools libs=shared,static arch=linux-rhel8-zen ^[deptypes=build,link virtuals=zlib-api] zlib-ng@=2.2.1%gcc@=8.5.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-rhel8-zen 

truncate: babelstream@=5.0%gcc@=8.5.0~acc+cuda~hip~ipo~kokkos~ocl~omp~raja~rocm~std~sycl~sycl2020~tbb+thrust 

list: ['cuda', 'thrust'] 

```
So Split makes sure that we only get the model variants

becker33 replied:
```
model_names = ["cuda", "std", ...]  # all model names
model_list = [name for name in model_names if f"+{name}" in self.spec]
```

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -235,183 +464,408 @@ def cmake_args(self):
+            # ------------
+            DEVICE = "-DUSE_DEVICE -cuda -gpu=nomanaged"
+            MANAGED = ""
+            config["FCFLAGS"] = flags
```
All the `flags` logic would be best handled in a flag_handler method so that it can incorporate the flags from `fflags=...` on the spec.

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -235,183 +464,408 @@ def cmake_args(self):
+        #               FJ
+        # ===================================
+        if spec.compiler.name == "fj":
+            flags = "-X08 -Kfast -KA64FX -KSVE -KARMV8_3_A -Kzfill=100 "
+            flags += "-Kprefetch_sequential=soft "
+            flags += "-Kprefetch_line=8 -Kprefetch_line_L2=16 -Koptmsg=2 "
+            # FJ Fortran system_clock is low resolution
+            flags += "-Keval -DUSE_OMP_GET_WTIME=1 "
+
+            config["DOCONCURRENT_FLAG"] = "-Kparallel,reduction -DNOTSHARED"
+            config["ARRAY_FLAG"] = "-Kparallel,reduction"
+            config["OPENMP_FLAG"] = pkg.compiler.openmp_flag
```
Am I missing something? It doesn't look like this `flags` variable is ever used.

kaanolgu replied:
You are right, it is missing from some of the fortran models adding them too :)

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -235,183 +464,408 @@ def cmake_args(self):
+        #               NVHPC
+        # ===================================
+        if spec.compiler.name == "nvhpc":
+            flags = pkg.compiler.opt_flags[4]  # for -O3
```
It looks like this variable is set for every compiler, and only used for the `nvhpc` compiler.

kaanolgu replied:
Thanks yes will be updating as I said in previous comment, thanks again!

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -235,183 +464,408 @@ def cmake_args(self):
+            spec_target = "znver3" if str(spec.target) == "zen3" else str(spec.target)
+            flags += "-march=" + spec_target
+
+            config["FCFLAGS"] = flags
```
flags is also used for this compiler.

kaanolgu replied:
Will rename it as `fortran_flags`

kaanolgu replied:
Would it be wise to set `fortran_flags` inside of each model but set 

```
config["FCFLAGS"] = flags
``` 
Outside of these if conditions only once instead of every model separately? I think that was my aim but missed a point

becker33 replied:
Resolved in 26ccd

Commit:
[fix] flags variant, flags for fortran set properly

Commit:
[fix] deleting excess =

kaanolgu committed:
Merge branch 'develop' of https://github.com/spack/spack into develop

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -108,310 +218,657 @@ class Babelstream(CMakePackage, CudaPackage, ROCmPackage):
+                        offload_args = " -mp=gpu;" + "-gpu=" + cuda_arch + " "
+                    if self.spec.satisfies("%clang"):
+                        cuda_arch = "sm_" + self.spec.variants["cuda_arch"].value[0]
+                        offload_args = "-march=znver3;-fopenmp;--offload-arch=" + cuda_arch
```
```suggestion
                        offload_args = "-fopenmp;--offload-arch=" + cuda_arch
```

kaanolgu committed:
Update var/spack/repos/builtin/packages/babelstream/package.py

Co-authored-by: Greg Becker <becker33@llnl.gov>

kaanolgu committed:
Merge branch 'develop' of github.com:kaanolgu/spack into develop

kaanolgu committed:
simplifying the complex logic of model pick

kaanolgu committed:
changing model dependency to be languages rather than build system

kaanolgu committed:
removing hardcoded arch flags and replacing with archspec

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -5,100 +5,217 @@
-    variant("cpu_arch", values=str, default="none", description="Enable CPU Target for ACC")
-    variant("acc_target", values=str, default="none", description="Enable CPU Target for ACC")
+
+    variant(
```
Couldn't we just use the spec target for this instead of an additional variant? It's a bit weird to say `babelstream cpu_arch=zen3 target=zen4`.

kaanolgu replied:
Addressed in the new commit [d6400d4](https://github.com/spack/spack/pull/41019/commits/d6400d437dbfd05801f969e43eeb1ef36abbf266) Thanks!

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -5,100 +5,217 @@
-
+    maintainers("tomdeakin", "kaanolgu", "tom91136")
+    # Previous maintainers: "robj0nes"
+    depends_on("cxx", type="build")
```
```suggestion
    depends_on("cxx", type="build", when="languages=cxx")
    depends_on("fortran", type="build", when="languages=fortran")
```

becker33 commented on var/spack/repos/builtin/packages/babelstream/package.py:
```diff
@@ -5,100 +5,217 @@
-    variant("stdindices", default=False, description="Enable STD-indices support")
-    variant("stdranges", default=False, description="Enable STD-ranges support")
+    # in the future it could be possible to add other languages too
+    variant(
+        "languages",
+        default="cxx",
+        values=("cxx", "fortran"),
+        description="Languages Babelstream Spack Package Support"
+        )
```
This implies that it's one or the other, but never both. Is that correct?

kaanolgu replied:
Yes

kaanolgu committed:
Update var/spack/repos/builtin/packages/babelstream/package.py

Co-authored-by: Greg Becker <becker33@llnl.gov>

kaanolgu committed:
Merge branch 'develop' of https://github.com/spack/spack into develop

kaanolgu committed:
removing cpu_arch from acc model

kaanolgu committed:
Merge branch 'develop' of github.com:kaanolgu/spack into develop

kaanolgu commented:
@spackbot fix style

kaanolgu committed:
[@spackbot] updating style on behalf of kaanolgu

kaanolgu commented:
@spackbot fix style

kaanolgu committed:
fixing style removing re

kaanolgu committed:
fixing style, removing long comment

kaanolgu committed:
Merge branch 'develop' into develop
